BFloat16: add explicit dtype support for to_mkldnn and to_dense

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37218 BFloat16: enable prepacked weights's inference
* **#37215 BFloat16: add explicit dtype support for to_mkldnn and to_dense**
* #37199 add mkldnn relu backward and reshape backward
* #37147 add mkldnn batch_norm backward
* #37146 add mkldnn pooling backward
* #36122 add mkldnn linear backward
* #36121 enable mkldnn conv2d backward to support mkldnn tensor input

Differential Revision: [D22440964](https://our.internmc.facebook.com/intern/diff/D22440964)

XiaobingSuper committed:
BFloat16: add explicit dtype support for to_mkldnn and to_dense

[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"




[ghstack-poisoned]

XiaobingSuper commented:
@jgong5, @hongzhen1

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"




[ghstack-poisoned]

albanD commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -337,7 +337,7 @@ SparseTensor dense_to_sparse(const Tensor& self, int64_t sparse_dim){
 // NB: Dropped the resizeNd variants
 
-Tensor sparse_to_dense(const SparseTensor& self) {
+Tensor sparse_to_dense(const SparseTensor& self, c10::optional<ScalarType> dtype) {
```
The new flag is completely ignored for here. Is that expected?

XiaobingSuper replied:
This is just for dispatch which should have same numbers of inputs.

VitalyFedyunin replied:
In this case please error out if dtype is provided, as we prefer not to have situations when user passing ignored argument.

albanD commented on aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp:
```diff
@@ -8,35 +8,53 @@ namespace at { namespace native {
-  AT_ASSERTM(cpu_tensor.dim() <= 5,
-             "Can't convert cpu tensor with the number of dimensions > 5");
+Tensor dense_to_mkldnn(const Tensor& cpu_tensor, c10::optional<ScalarType> dtype) {
+  TORCH_INTERNAL_ASSERT(cpu_tensor.device().type() == DeviceType::CPU,
```
Why are these internal asserts? These args are not controled by the user? If so, you prefer a TORCH_CHECK here I think.

albanD commented on aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp:
```diff
@@ -8,35 +8,53 @@ namespace at { namespace native {
-                    ideep::tensor::data_type::f32,
-                    (cpu_tensor_cont.template data_ptr<float>()));
+  if (cpu_tensor.scalar_type() == ScalarType::Float) {
+    dtensor.feed_from(dtensor.get_dims(),
```
This function will work for any `dtype` provided?

XiaobingSuper replied:
Yes, the default dtype is float, there can support othe dtype, such as int8.

albanD commented on aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp:
```diff
@@ -8,35 +8,53 @@ namespace at { namespace native {
   if (stensor.is_empty()) return cpu_tensor;
-  auto pub_tensor = stensor.to_public(cpu_tensor.template data_ptr<float>());
+  auto pub_tensor =
+      cpu_tensor.scalar_type() == ScalarType::Float
```
Here the cpu_tensor type is dtype. So it can be other things that float/bfloat right?

XiaobingSuper replied:
yes.

albanD reviewed (changes requested):
How general is this supposed to be? If there are restrictions on the dtype, a couple checks are missing I think.

Also for sparse Tensor, if the output dtype arg is not supported, and error should be raised if the user passes it.

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"


Differential Revision: [D22440964](https://our.internmc.facebook.com/intern/diff/D22440964)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: add explicit dtype support for to_mkldnn and to_dense"


Differential Revision: [D22440964](https://our.internmc.facebook.com/intern/diff/D22440964)

[ghstack-poisoned]

VitalyFedyunin reviewed (commented):
Several error messaging fixes required, everything else looks good.

VitalyFedyunin commented:
Please rebase entire stack to make it mergable.
